core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -566,6 +574,13 @@ private[spark] class ExecutorAllocationManager(
     if (executorIds.contains(executorId)) {
       executorIds.remove(executorId)
       removeTimes.remove(executorId)
+      if (!removedExecutorIds.offer(executorId)) {
+        logWarning(s"The queue is full for tracking removed executors, will trim old " +
+          s"executor ids and offer again")
+        val trimSize = math.max(1, removedExecutorIds.size() / 10)
+        (0 to trimSize).foreach(removedExecutorIds.poll())
+        removedExecutorIds.offer(executorId)
```

Comment:
Can we ensure that this `offer` always succeed?  If `removedExecutorIds` uses `LinkedBlockingQueue` for multi-threading, line 577 is executed on multi-threads while line 581 is running. In that case, line 582 may see full queue again.

Reply:
@kiszk  Yes, I think `ArrayDeque` is ok, and I was thinking if  there is a better solution not based on the following assumption, any other good solution for this?

>  // Assumes the late events for the executor will be correctly handled before the removal

Reply:
or how about a LinkedHashSet?   Below you are doing a `removedExecutorIds.contains(executorId)` which will be slow with 10K elements.

Reply:
@squito Good suggestion!